src/backend/distributed/planner/multi_join_order.c

```diff
@@ -905,6 +941,8 @@ SinglePartitionJoin(JoinOrderNode *currentJoinNode, TableEntry *candidateTable,
 	/*
 	 * If we previously dual-hash re-partitioned the tables for a join or made
 	 * cartesian product, we currently don't allow a single-repartition join.
+	 *
+	 * TODO: verify what happens when there was an other join in between our current evaluation and below joins, this pattern has caused crashes before.
```

Comment:
can this todo be resolved?

Reply:
I now have the knowledge of what kind of query to create to trigger the case that would verify its behaviour. I will add it to the test suite and remove the TODO

Reply:
Removed the TODO.

There is a bit of weird behaviour but not crashing, there are queries in our test suite covering this.

The side effect of `DUAL_PARTITION_JOIN` is that it loses the `currentPartitionColumn` since there is no common column after a dual repartition. `SinlgePartitionJoinClause` will always return `NULL` when there is not `currentPartitionColumn`. It could still enter a single partition clause  afterwards if it could partition to the candidate column.

`CARTESIAN_PRODUCT` strangely retains its current partition column which sounds like something it should not, but given the cartesian product will never execute it doesn't matter much. I wonder if we at some point shoulld remove the cartesian join order planner from the planner as it will never end up with an executable path, and the error message is quite vague.